doc/src/manual/arrays.md

```diff
@@ -1028,33 +1028,57 @@ It is sometimes useful to perform element-by-element binary operations on arrays
 sizes, such as adding a vector to each column of a matrix. An inefficient way to do this would
 be to replicate the vector to the size of the matrix:
 
-```julia-repl
-julia> a = rand(2,1); A = rand(2,3);
+```jldoctest broadcasting
+julia> a = [1, 2]; B = [10 20 30; 40 50 60]
 
-julia> repeat(a,1,3)+A
-2×3 Array{Float64,2}:
- 1.20813  1.82068  1.25387
- 1.56851  1.86401  1.67846
+julia> repeat(a, 1, 3) + B
+2×3 Matrix{Int64}:
+ 11  21  31
+ 42  52  62
 ```
 
-This is wasteful when dimensions get large, so Julia provides [`broadcast`](@ref), which expands
-singleton dimensions in array arguments to match the corresponding dimension in the other array
-without using extra memory, and applies the given function elementwise:
+This is wasteful when dimensions get large, so Julia provides
+[`broadcast`](@ref), which aligns dimensions in array arguments starting with
+the leading ones[^2], fills lacking dimensions with a singleton dimension,
+expands singleton dimensions to match the corresponding dimension in the other
+array without using extra memory, and applies the given function elementwise:
 
-```julia-repl
-julia> broadcast(+, a, A)
-2×3 Array{Float64,2}:
- 1.20813  1.82068  1.25387
- 1.56851  1.86401  1.67846
-
-julia> b = rand(1,2)
-1×2 Array{Float64,2}:
- 0.867535  0.00457906
-
-julia> broadcast(+, a, b)
-2×2 Array{Float64,2}:
- 1.71056  0.847604
- 1.73659  0.873631
+[^2]: This is the opposite direction of that of NumPy, which aligns dimensions
+from the last to the first.
+
+```jldoctest broadcasting
+julia> broadcast(+, a, B)
+2×3 Matrix{Int64}:
+ 11  21  31
+ 42  52  62
+
+julia> c = [10 20 30]
+1×3 Matrix{Int64}:
+ 10  20  30
+
+julia> broadcast(+, a, c)
+2×3 Matrix{Int64}:
+ 11  21  31
+ 12  22  32
+```
+
+In the example above, the [`size`](@ref)s of array `a` and `B` are `(2,)` and
+`(2, 3)`, respectively. Therefore, broadcasting `a` and `B` fills the second
+dimension of `a` with one and then expands it to three so that the two arrays
+have the same size.  Similarly, since the size of `c` is `(1, 2)`, broadcasting
+`a` and `c` fills the second dimension of `a` with one and expands the second
+dimension of `a` to three and the first dimension of `c` to two.
+
+```
+Broadcasting of a and B:
+           fill         expand
```

Comment:
Here (and above) you say "fill" and "expand". But `fill` means something else. Perhaps clearer to use always the verbs which are commands: "reshape" and "repeat".

I also think the paragraph above could be much shorter. It need not read the whole table out in prose before showing the table. Just show the table, find one sentence to introduce it?

I'd also make `c` not re-use the numbers from `B`, so that the results are visually more distinct. Maybe it should even be `c = [10 100 1000;;;]` so that the example also introduces dimensions beyond the 2nd, even if they are trivial.

Reply:
Hmm, I think "fill" is less confusing because it is obvious from the context that it is used as a general English word here. Using "reshape" or "repeat" might be more confusing because functions with these names exist in Julia and the reader might misunderstand that `reshape` or `repeat` is called when broadcasting arrays.

The paragraph might look verbose but I think we should provide enough explanation to introduce a new concept. While the behavior of broadcasting is almost obvious to us, it is not always the case to others.  I think this paragraph is important for readers to confirm their mental model in words.

I've changed `c` to `[10 100 1000]`. `[10 100 1000;;;]` looks distracting to me because the syntax is rarely used.